### ISSUE TITLE
How to get complete data in Twitter API Premium?

### ISSUE THREAD

**Issue by ardiantovn** [geduldig/TwitterAPI#136]
I want to get data from March 8 2020 to April 5 2020. I do crawling with the program code as below. After completing data retrieval, I only get data the last few days in April. Why didn't I get the data in March? Which part do I need to fix from this program code?

My code:
```
import json as jsond
from TwitterAPI import TwitterAPI

PRODUCT= 'fullarchive'
LABEL = 'mylabel'
SEARCH_TERM = 'covid19 OR donasi'
topic = 'corona'
count= 0
maxcount = 5000

api = TwitterAPI(
             consumer_key, 
             consumer_secret, 
             access_token_key, 
             access_token_secret)
nextE = ''
while True:
    if count==0:
        r = api.request('tweets/search/%s/:%s' % (PRODUCT, LABEL), 
                        {'query'     :SEARCH_TERM, 
                         #'maxResults':'500',
                         'fromDate' :'202003080000',
                         'toDate'   :'202004052359'
                        })
    else:
        r = api.request('tweets/search/%s/:%s' % (PRODUCT, LABEL), 
                        {'query'     :SEARCH_TERM, 
                         'maxResults':'500',
                         'fromDate' :'202003080000',
                         'toDate'   :'202004052359',
                         'next'      :nextE
                        })
    if r.status_code != 200:
        print(r.status_code)
        break
    for i in r.json()['results']:
        i['topic']=topic
        jsond.dump(str(i),fw)
        fw.write("\n")
        count = count + 1
    json_resp = r.json()
    print(json_resp['next'])
    if 'next' not in json_resp:
        break
    nextE = json_resp['next']
    
    print('GOT :%d'%(count))
    if count >= maxcount:
        break
    time.sleep(2)
fw.close()
```

Thank you

**Comment by geduldig** [comment]
I'm curious how many results you actually get in the last couple of days.

Are you in a Premium sandboxed environment? Sandboxes are limited to 100 max results per request.

Are you getting a thrown exception? One thing you are missing is exception handling which might point you to the problem. At a mininum you should print the status code, but Twitter will also return error messages.

Do you  know where in the `while` loop you are exiting?

Are you sure you are not exceeding the rate limit?  I see that after each request you are sleeping 2 sec. This may be cutting it too close to the rate limit of 30 requests per minute.
 
Lastly, after you figure out what is the problem, you may want to look at `TwitterPager` (described [here](https://geduldig.github.io/TwitterAPI/paging.html). This class is part of `TwitterAPI` and takes care of `next` and rate limits.

**Comment by ardiantovn** [comment]
-I get the last 3 days(April 3 2020 until April 5 2020).
-I try Premium and Premium Sandboxed API. 
-Yes, I sometimes get error limiting
-The while loop exit when count>=Maxcount.

In the past, I also tried using TwitterPager.
I tried to install a range of data for example January 1, 2020 to January 31, 2020. As I recall (I did not save the data), the data was only taken for the last few days (for example, January 26, 2020, January 31, 2020). I was unable to retrieve further data because I ran out of the request quota. Is there a way to do data sampling? I hope I can sample data for example 10% of the total number of tweets every day.

**Comment by geduldig** [comment]
> -I get the last 3 days(April 3 2020 until April 5 2020).
I was asking *how many* tweets you got the past couple of days.

> -I try Premium and Premium Sandboxed API.
Change 'maxResults':'500', to 100.

> -Yes, I sometimes get error limiting
Change time.sleep(2) to 3.

> -The while loop exit when count>=Maxcount.
This is why you only get a couple of days of tweets. You have reached maxcount.

> 
> In the past, I also tried using TwitterPager.
> I tried to install a range of data for example January 1, 2020 to January 31, 2020. As I recall (I did not save the data), the data was only taken for the last few days (for example, January 26, 2020, January 31, 2020). I was unable to retrieve further data because I ran out of the request quota. Is there a way to do data sampling? I hope I can sample data for example 10% of the total number of tweets every day.
What is "data sampling"? You can try the Twitter's streaming endpoint `filter`. However, this samples only current tweets in realtime, not past tweets.

**Comment by ardiantovn** [comment]
I am sorry, I can't remember the amount of data. I did not save the data.

Thank you for answering the second to fourth questions.

The sampling data I mean is like this:
On March 1 2020 00:00 to March 1 2020 23:59 there are 1000 tweets. I just want to take 100 tweets randomly out of the total tweets. After that I continued to retrieve data the same way the other day.

**Comment by ardiantovn** [comment]
I want to ask to confirm the knowledge that I understand.

Suppose I set the value `begindate = 202001010000` and `enddate = 202001312359`. Apart from that, I don't use `maxCount` to stop the while loop. The loop will stop when the request quota runs out. Then the TwitterAPI will retrieve data backwards from 31 January 2020 23:59 to 1 January 2020 00:00. Is that true?
If so, it could be that I ran out of the quota before all data was collected because there were so many tweets available. It could be, I will only get data from January 15, 2020 to January 31, 2020. Is my statement true?
